frontend/mux.go

```diff
@@ -0,0 +1,515 @@
+}
+
+func (d *clientWithCustomOpts) BuildOpts() gwclient.BuildOpts {
+	return d.opts
+}
+func (d *clientWithCustomOpts) CurrentFrontend() (*llb.State, error) {
+	return d.Client.(CurrentFrontend).CurrentFrontend()
+}
+
+// Handler returns a [gwclient.BuildFunc] that uses the mux to route requests to appropriate handlers
+func (m *BuildMux) Handler(opts ...func(context.Context, gwclient.Client, *BuildMux) error) gwclient.BuildFunc {
+	return func(ctx context.Context, client gwclient.Client) (*gwclient.Result, error) {
+		if !SupportsDiffMerge(client) {
+			dalec.DisableDiffMerge(true)
+		}
+		for _, opt := range opts {
+			if err := opt(ctx, client, m); err != nil {
+				return nil, err
+			}
+		}
+		return m.Handle(ctx, client)
+	}
+}
+
+// WithTargetForwardingHandler registers a handler for each spec target that has a custom frontend
+func WithTargetForwardingHandler(ctx context.Context, client gwclient.Client, m *BuildMux) error {
+	if k := GetTargetKey(client); k != "" {
+		// This is already a forwarded request, so we don't want to forward again
+		return fmt.Errorf("target forwarding requested but target is already forwarded: this is a bug in the frontend for %q", k)
+	}
+	spec, err := m.loadSpec(ctx, client)
+	if err != nil {
+		return err
+	}
+
+	for key, t := range spec.Targets {
+		if t.Frontend == nil {
+			continue
+		}
+
+		m.Add(key, func(ctx context.Context, client gwclient.Client) (*gwclient.Result, error) {
+			ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithField("frontend", key).WithField("frontend-ref", t.Frontend.Image).WithField("forwarded", true))
+			bklog.G(ctx).Info("Forwarding to custom frontend")
+			req, err := newSolveRequest(
+				copyForForward(ctx, client),
+				withSpec(ctx, spec, dalec.ProgressGroup("prepare spec to forward to frontend")),
+				toFrontend(t.Frontend),
+				withTarget(client.BuildOpts().Opts[keyTarget]),
+			)
+
+			if err != nil {
+				return nil, err
+			}
+
+			return client.Solve(ctx, req)
+		}, nil)
+		bklog.G(ctx).WithField("target", key).WithField("targets", maps.Keys(m.handlers)).WithField("targetKey", GetTargetKey(client)).Info("Added custom frontend to router")
+	}
+	return nil
+}
+
+// WithBuiltinHandler registers a late-binding handler for the given target key.
+// These are only added if the target is in the spec OR the spec has no explicit targets.
+func WithBuiltinHandler(key string, bf gwclient.BuildFunc) func(context.Context, gwclient.Client, *BuildMux) error {
+	return func(ctx context.Context, client gwclient.Client, m *BuildMux) error {
+		spec, err := m.loadSpec(ctx, client)
+		if err != nil {
+			return err
+		}
+
+		if len(spec.Targets) > 0 {
+			t, ok := spec.Targets[key]
+			if !ok {
+				bklog.G(ctx).WithField("spec targets", maps.Keys(spec.Targets)).WithField("targetKey", key).Info("Target not in the spec, skipping")
+				return nil
+			}
+
+			if t.Frontend != nil {
+				bklog.G(ctx).WithField("targetKey", key).Info("Target has custom frontend, skipping builtin-handler")
```

Comment:
there is the mention that has a custom frontend and is going to be skipped but only seems to be log it? does it need a return nil?

Reply:
Fixed. Good catch!